_posts/2024-01-19-trino-2023-wrapped.md

```diff
@@ -0,0 +1,253 @@
+order to create and re-use functions that do exactly what you need them to do.
+
+If you want to learn more about SQL routines, you can check
+out [the introduction to SQL routines](/docs/current/routines/introduction.html)
+in our documentation, as well as a
+few [example routines](/docs/current/routines/examples.html) which give a good
+look at how they can be used.
+
+### Schema evolution and dynamic catalogs
+
+While we’re providing more power, customization, and flexibility to Trino users,
+it’s also important to highlight just how much has been added this year to make
+it easier to adjust things on the fly.
+
+Schema evolution came first, allowing you to alter columns’ data types, rename
+columns, and handle nested fields when dropping columns. Instead of needing to
+use the underlying database or modify it some other way and reboot Trino, Trino
+is now aware of when your schema changes and can handle the adjustments on the
+fly.
+
+But we’ve experimentally taken things one step further in 2023, adding dynamic
+catalogs to Trino. Rather than adjusting your schema one column at a time, what
+about adding or dropping an entire catalog in one go? You can do that now.
+Though it’s currently still bleeding-edge and not ready for widespread use on
+your important production tables, we’re looking forward to improving it and
+making it resilient and stable in 2024.
+
+### Project Hummingbird
+
+Trino has always been about squeezing out every ounce of performance that you
+can get. You can go check out our [release notes](/current/release.html) and
+you’ll see that every version has at least a couple performance improvements in
+it, if not more. Over time, these performance improvements add up to a
+substantial gain, meaning that version-over-version, year-over-year, Trino is
+always getting faster. Project Hummingbird was a concerted effort this year to
+take a look at the core engine and make a number of small improvements that
+would add up to something
+substantial. [The GitHub issue tracking it](https://github.com/trinodb/trino/issues/14237)
+is gloriously nearly done, with a lot of that work done in 2023.
+
+### Lakehouse improvements
+
+Want to leverage the historical log of all actions taken on a table in Hudi? The
+new `$timeline` system table has you covered. How about in Delta Lake? We’ve got
+the `table_changes` function for that, and views were added there, too. Too many
+metadata tables to list were added to Iceberg, along with the REST, JDBC, and
+Nessie catalogs for metadata.
+
+### Java 21?
+
+Java 21. It’s required to run the latest version of Trino now. With the upgrade
+from Java 17 to 21 comes a ton of improvements that will make development on
+Trino easier and better than ever, which will in turn make it faster and
+smoother than ever. Though not as huge of a deal as our upgrade to Java 17 last
+year, expect to see the benefits coming down the pipeline as the engineers
```

Comment:
I dont think it was last year we switched to 17 .. feels like longer ago

Reply:
July 2022